db/src/db.rs

```diff
@@ -285,60 +283,62 @@ pub fn ensure_current_version(conn: &mut rusqlite::Connection) -> Result<i32> {
-            Value::Integer(x) => Ok(ToSqlOutput::from(x)),
-            // TODO: consider using a larger radix to save characters.
-            Value::BigInteger(ref x) => Ok(ToSqlOutput::from(x.to_str_radix(10))),
-            Value::Float(ref x) => Ok(ToSqlOutput::from(x.into_inner())),
-            // TODO: try to avoid this clone.
-            Value::Text(ref x) => Ok(ToSqlOutput::from(x.clone())),
-            Value::PlainSymbol(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::NamespacedSymbol(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::Keyword(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::NamespacedKeyword(ref x) => Ok(ToSqlOutput::from(x.to_string())),
-            Value::Vector(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
-            Value::List(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
-            Value::Set(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
-            Value::Map(_) => Err(rusqlite::Error::InvalidColumnName(format!("Cannot convert to_sql: {:?}", self.0))),
+    /// Given an EDN `value`, return a corresponding Mentat `TypedValue`.
+    ///
+    /// An EDN `Value` does not encode a unique Mentat `ValueType`, so the composition
+    /// `from_edn_value(first(to_edn_value_pair(...)))` loses information.  Additionally, there are
+    /// EDN values which are not Mentat typed values.
+    ///
+    /// This function is deterministic.
+    pub fn from_edn_value(value: &Value) -> Option<TypedValue> {
+        match value {
+            &Value::Boolean(x) => Some(TypedValue::Boolean(x)),
+            &Value::Integer(x) => Some(TypedValue::Long(x)),
+            &Value::Float(ref x) => Some(TypedValue::Double(x.clone())),
+            &Value::Text(ref x) => Some(TypedValue::String(x.clone())),
+            &Value::NamespacedKeyword(ref x) => Some(TypedValue::Keyword(x.to_string())),
+            _ => None
+        }
+    }
+
+    /// Return the corresponding SQLite `value` and `value_type_tag` pair.
+    pub fn to_sql_value_pair<'a>(&'a self) -> (ToSqlOutput<'a>, i32) {
+        match self {
+            &TypedValue::Ref(x) => (rusqlite::types::Value::Integer(x).into(), 0),
+            &TypedValue::Boolean(x) => (rusqlite::types::Value::Integer(if x { 1 } else { 0 }).into(), 1),
+            // SQLite distinguishes integral from decimal types, allowing long and double to share a tag.
+            &TypedValue::Long(x) => (rusqlite::types::Value::Integer(x).into(), 5),
+            &TypedValue::Double(x) => (rusqlite::types::Value::Real(x.into_inner()).into(), 5),
+            &TypedValue::String(ref x) => (rusqlite::types::ValueRef::Text(x.as_str()).into(), 10),
+            &TypedValue::Keyword(ref x) => (rusqlite::types::ValueRef::Text(x.as_str()).into(), 13),
```

Comment:
By keeping a `String` in the `Keyword`, this doesn't copy data until it hits SQLite.  If we keep a `Keyword` (really a `NamespacedKeyword`, since I'm being strict in my parser for now) then we have to cons strings and can't use the `ValueRef` in this way.